R/standardize_treatment_kernel.R

```diff
@@ -0,0 +1,448 @@
+
+}
+
+#' Compute block diagonal kernel matrix
+#' @param Xs List of covariate matrices by site
+#' @param kernel Kernel to use
+compute_block_diag_kernel <- function(Xs, kernel) {
+
+  # block diagonal kernel matrix
+  kern_mat <- Matrix::bdiag(lapply(Xs, function(x) kernlab::kernelMatrix(kernel, x)))
+  return(kern_mat)
+
+}
+
+#' Create diagonal regularization matrix
+#' @param pro_trt_split List of J vectors of treatment propensity multipliers
+#' @param pro_ctr_split List of J vectors of control propensity multipliers
+#' @param scale_sample_size Whether to scale the dispersion penalty by the sample size of each group, default T
+#' @param nj Number of units in each group
+#' @param n Total number of units
+create_I0_matrix_treatment_kernel <- function(pro_trt_split, pro_ctr_split, scale_sample_size, nj, n) {
+
+  if(scale_sample_size) {
+    # diagonal matrix of inverse propensity scores scaled by group sample size
+    I0 <- Matrix::Diagonal(n, rep(nj ^ 2, nj) * (unlist(pro_trt_split) + unlist(pro_ctr_split)))
+  } else {
+    # diagonal matrix of inverse propensity scores not scaled by group sample size
+    I0 <- Matrix::Diagonal(n, rep(nj, nj) * (unlist(pro_trt_split) + unlist(pro_ctr_split)))
+  }
+  return(I0)
+}
+
+#' Create the q vector for an QP that solves min_x 0.5 * x'Px + q'x
+#' @param Xtaus List of J n x dtau matrices of covariates split by group
+#' @param pro_trt_split List of J vectors of treatment propensity multipliers
+#' @param target Vector of population means to re-weight to
+#'
+#' @return q vector
+create_q_vector_treatment_kernel <- function(Xtaus, pro_trt_split, target) {
+  q <- -c(do.call(rbind, Xtaus) %*% target) * unlist(pro_trt_split)
+  return(q)
+}
+
+
+#' Create the P matrix for an QP that solves min_x 0.5 * x'Px + q'x
+#' @param n Total number of units
+#' @param X0s List of J n x d0 matrices of covariates split by group
+#' @param Xtaus List of J n x dtau matrices of covariates split by group
+#' @param kernel0 Kernel to use for average control potential outcome function
+#' @param kerneltau Kernel to use for average treatment effect function
+#' @param pro_trt_split List of J vectors of treatment propensity multipliers
+#' @param pro_ctr_split List of J vectors of control propensity multipliers
+#'
+#' @return P matrix
+create_P_matrix_treatment_kernel <- function(n, X0s, Xtaus, kernel0, kerneltau,
+                                             pro_trt_split, pro_ctr_split) {
+
+  # construct kernel matrices
+  kern_mat0 <- compute_block_diag_kernel(X0s, kernel0)
+  kern_mattau <- compute_block_diag_kernel(Xtaus, kerneltau)
+
+  # construct propensity matrices
+  pro_diff_unsplit <- unlist(pro_trt_split) - unlist(pro_ctr_split)
+  pro_mat0 <- pro_diff_unsplit %*% t(pro_diff_unsplit)
+  pro_trt_unsplit <- unlist(pro_trt_split)
+  pro_mattau <- pro_trt_unsplit %*% t(pro_trt_unsplit)
+
+  # combine matrices
+  P <- (kern_mat0 * pro_mat0) + (kern_mattau * pro_mattau)
```

Comment:
I don't really follow what's going on here.

Reply:
I'll DM you.